lib/audited/rspec_matchers.rb

```diff
@@ -109,31 +106,63 @@ def associated_with_model?
       end
 
       def records_changes_to_specified_fields?
-        if @options[:only] || @options[:except]
-          if @options[:only]
-            except = model_class.column_names - @options[:only].map(&:to_s)
-          else
-            except = model_class.default_ignored_attributes + Audited.ignored_attributes
-            except |= @options[:except].collect(&:to_s) if @options[:except]
-          end
+        ignored_fields = build_ignored_fields_from_options
 
-          expects "non audited columns (#{model_class.non_audited_columns.inspect}) to match (#{except})"
-          model_class.non_audited_columns =~ except
-        else
-          true
-        end
+        expects "non audited columns (#{model_class.non_audited_columns.inspect}) to match (#{ignored_fields})"
+        model_class.non_audited_columns.to_set == ignored_fields.to_set
       end
 
       def comment_required_valid?
-        if @options[:comment_required]
-          @subject.audit_comment = nil
+        expects "to require audit_comment before #{model_class.audited_options[:on]} when comment required"
+        requires_comment_before_callbacks?
+      end
+
+      def only_audit_on_designated_callbacks?
+        {
+          create: [:after, :audit_create],
+          update: [:before, :audit_update],
+          destroy: [:before, :audit_destroy]
+        }.map do |(action, kind_callback)|
+          kind, callback = kind_callback
+          callbacks_for(action, kind: kind).include?(callback) if @options[:on].include?(action)
+        end.compact.all?
+      end
+
+      def requires_comment_before_callbacks?
+        [:create, :update, :destroy].map do |action|
+          if @options[:comment_required] && model_class.audited_options[:on].include?(action)
+            callbacks_for(action).include?(:require_comment)
+          end
+        end.compact.all?
+      end
+
+      def callbacks_for(action, kind: :before)
+        model_class.send("_#{action}_callbacks").select { |cb| cb.kind == kind }.map(&:filter)
+      end
 
-          expects "to be invalid when audit_comment is not specified"
-          @subject.valid? == false && @subject.errors.key?(:audit_comment)
+      def build_ignored_fields_from_options
```

Comment:
This logic originally lived in `records_changes_to_specified_fields?`. I wrote some comments to myself on that method and this one which I share below:


```rb
# second branch of the case statement was conditionally unioning `except`
# if the `except` option was present, otherwise it would just return default
# ignored fields; while logically correct, I thought it might be clearer
# if we were explicit about the cases in wich the ignored attributes were build
# i.e. either when `only` is present, `except` is present, or neither is present
# I also changed `except` to `ignored_fields` because I found it a bit confusing
# to be using a variable name that matched one of the options to the auditor
# even though the values within the array weren't necessarily the values
# passed into the option array

# I also got rid of the default `true` because even if neither option is passed in
# we should probably check and see that the default ignored fields are still
# not being audited
```